Add Prism AST node factory

allcre opened the pull request:
Part of #9065 

This PR introduces a `parser::Prism::Factory` class that provides high-level helpers for constructing Prism AST nodes, mirroring the functionality of `ast::MK` helpers but for raw Prism nodes (`pm_node_t*`)

### Test plan
<!-- If you did not write tests for this change, replace the message below explaining why not. Why we should be confident this change is correct? If you changed the website, please include a screenshot of the proposed changes. -->

Will be covered by existing tests once used in RBS rewriting

allcre committed:
Add Prism AST node factory

Co-authored-by: Alexander Momchilov <alexander.momchilov@shopify.com>
Co-authored-by: Jesse Johnson <jesse.johnson@shopify.com>
Co-authored-by: Kaan Ozkan <kaan.ozkan@shopify.com>

jez reviewed (dismissed):
Question about whether we can restructure this a little bit

froydnj commented on parser/prism/Factory.cc:
```diff
@@ -0,0 +1,512 @@
+// * In the event of an exception, will be freed correctly using Prism's `xfree()`.
+template <typename T> using PrismUniquePtr = std::unique_ptr<T, decltype(prismFree)>;
+
+pm_node_list_t Factory::copyNodesToList(const vector<pm_node_t *> &nodes) const {
```
Since this function doesn't really care about the vector-ness of the list, it should take an `absl::Span<pm_node_t *>`, so the caller isn't forced to allocate a vector.

froydnj commented on parser/prism/Factory.cc:
```diff
@@ -0,0 +1,512 @@
+    return up_cast(createSendNode(T(loc), methodId, up_cast(arguments), tinyLoc, fullLoc, tinyLoc));
+}
+
+pm_node_t *Factory::TAll(core::LocOffsets loc, const vector<pm_node_t *> &args) const {
```
`absl::Span` here and elsewhere in these higher-level constructs.

froydnj commented on parser/prism/Factory.h:
```diff
@@ -0,0 +1,109 @@
+
+    template <typename T> T *allocateNode() const {
+        return new T{};
+    }
```
This should be allocating things with Prism's allocator, not the general C++ one.

froydnj commented on parser/prism/Factory.cc:
```diff
@@ -0,0 +1,512 @@
+}
+
+void Factory::free(void *ptr) const { // see Prism's `include/prism/defines.h`
+    ENFORCE(ptr);
```
Does Prism's `xfree` really require this?  It looks like Prism defaults to using `free`, and `free` is certainly OK with null pointers.

froydnj reviewed (changes requested):
Is there an explanation of what the endgame is here?  I gather through looking at the linked Shopify issue that this is for doing RBS rewrites on the prism AST instead of the Sorbet one...but there's no explanation of the why.  (Or perhaps this has been discussed elsewhere and I either missed it or I have forgotten about it?)

My biggest concern with this is how it locks Sorbet into a very particular way of using Prism.  ruby/prism#2888 has (admittedly long-term and bluesky-ish) ideas about how to clean up Prism's C API, and the idea about `pm_node_t` being an opaque pointer seems to clash with this idea that we can freely allocate Prism's internal C structures.   In a future world where Prism enforces that kind of API, there are significant challenges for Sorbet to upgrade Prism versions.

I realize that the Prism definitions are entirely public, but I think the spirit of the definitions is that they are only obtainable from parsing a ruby file, not for constructing your own tree?  Doing the latter is...unusual?

This entire file also has to have very specific knowledge about any Prism-internal invariants it has to uphold (the node counter cited below is a good example -- I think it is ridiculous that Prism has to keep track of this information, and I don't think it's relevant to Sorbet's purposes, but this approach is reaching into Prism internals by creating Prism nodes itself and therefore has to deal with this kind of thing).  This also creates barriers to upgrading and subtle mismatches that will not be fun to debug if they cause problems.

amomchilov commented on parser/prism/Factory.cc:
```diff
@@ -0,0 +1,512 @@
+    prismParser->node_id++;
+    uint32_t nodeId = prismParser->node_id;
+
+    return (pm_node_t){.type = type, .flags = 0, .node_id = nodeId, .location = loc};
```
There are some that definitely aren't "correct" (e.g. `newline`), but not in a way that makes a difference to the final desugar tree.

allcre committed:
Address comments

froydnj commented:
@amomchilov Yes, I would like to see that the Prism maintainers are comfortable with this direction and/or have broad agreement on how things are going to proceed going forward.

allcre committed:
Fix build errors

- Adds missing helper methods
- Minor const changes

amomchilov commented on parser/prism/Factory.cc:
```diff
@@ -98,7 +99,7 @@ pm_node_t *Factory::ConstantPathNode(core::LocOffsets loc, pm_node_t *parent, st
     ENFORCE(arg, "SingleArgumentNode: arg is required");
 
-    vector<pm_node_t *> args = {arg};
+    array<pm_node_t *, 1> args = {arg};
```
Turns out you can use template deduction in cases like this:

```
    auto args = array{arg};
```

froydnj commented on parser/prism/Factory.cc:
```diff
@@ -0,0 +1,512 @@
+
+pm_node_t Factory::initializeBaseNode(pm_node_type_t type, const pm_location_t loc) const {
+    pm_parser_t *prismParser = parser.getRawParserPointer();
+    prismParser->node_id++;
```
I think this will be resolved by Prism updates.

froydnj reviewed (changes requested):
It looks like there are some build failures.  A couple of comments below.

Note that the previous comments about performance remain unaddressed.

KaanOzkan commented on parser/prism/Factory.h:
```diff
@@ -0,0 +1,109 @@
+    pm_node_t *T_Enumerator(core::LocOffsets loc) const;
+    pm_node_t *T_Hash(core::LocOffsets loc) const;
+    pm_node_t *T_Set(core::LocOffsets loc) const;
+    pm_node_t *T_Range(core::LocOffsets loc) const;
```
Are you thinking of another class called something like `SorbetHelpers` that takes in a factory instance?

The inspiration for this class was `helper.h` where both node and sorbet specific constructs are combined. It makes the callsites simpler as everything is being called on the same member but we could create a new class and add it as a member to all of its users.

KaanOzkan commented on parser/prism/Factory.cc:
```diff
@@ -0,0 +1,512 @@
+    PrismUniquePtr<uint8_t> stable{reinterpret_cast<uint8_t *>(this->calloc(nameLen, sizeof(uint8_t))), prismFree};
+
+    memcpy(stable.get(), name.data(), nameLen);
+    pm_constant_id_t id = pm_constant_pool_insert_owned(&prismParser->constant_pool, stable.release(), nameLen);
```
Does this new API require a separate class or would you be okay with doing it through the `SorbetHelpers` hinted at https://github.com/sorbet/sorbet/pull/9667#discussion_r2548229410.

I'm thinking we could have `pm_constant_id_t` member variables for each sorbet construct that get created once in the constructor.

KaanOzkan committed:
Improve Factory API

- Rename Send*/createSendNode to Call*/createCallNode for clarity
- Fix const-correctness: use const Span<pm_node_t*> instead of
  Span<const pm_node_t*> to match Prism's C API and remove const_cast
- Add new helper methods: Nil(), String(), TSelfType(), TAnything(),
  TAttachedClass(), T_Boolean(), T_Module()
- Add ConstantReadNode/ConstantPathNode overloads taking pm_location_t
- Simplify array construction using C-style arrays instead of std::array

KaanOzkan committed:
Improve Factory::SymbolFromConstant

KaanOzkan committed:
Make calloc a template function
